### PR TITLE
Bonding share

### PR DESCRIPTION
**Pull request by zapaz**
**DONE**
* bonding contract : bondTokens and redeemShares functions
* change ERC20 bondingShare as ERC1155
* add totalSupply function to ERC1155
* implement durationMultiplier
* add rounding block number 
* add formulas in a UbiquityFormulas library
* separate BondingSetup before test
* tests

**TO DO**
* bonding contract : uGOV management

**NOT DONE**
* bonding contract : redeemAllShares (or to be called off chain calling multiple times redeemShares)

### CODE REVIEW

**Comment by zgorizzo69** [comment]
please fix lint warnings

File: contracts/Bonding.sol [hunk 1]
```diff
@@ -97,85 +106,138 @@ contract Bonding is CollectableDust {
...
+    /*
+        Desposit function with uAD-3CRV LP tokens (stableSwapMetaPoolAddress)
+     */
+    function bondTokens(uint256 _lpsAmount, uint256 _weeks)
+        public
+        returns (uint256 _id)
+    {
+        require(
+            1 <= _weeks && _weeks <= 520,
+            "Bonding: duration must be between 1 and 250 weeks"
```
Comment (zgorizzo69) [inline contracts/Bonding.sol:125]: 520 not 250

File: contracts/Bonding.sol [hunk 2]
```diff
@@ -3,24 +3,33 @@ pragma solidity ^0.8.3;
...
+    bytes public data = "";
 
     UbiquityAlgorithmicDollarManager public manager;
 
-    uint256 public constant TARGET_PRICE = 1 ether; // 3Crv has 18 decimals
+    uint256 public constant TARGET_PRICE = uint256(1 ether); // 3Crv has 18 decimals
     // Initially set at $1,000,000 to avoid interference with growth.
-    uint256 public maxBondingPrice = uint256(1 ether) * 1000000;
+    uint256 public maxBondingPrice = uint256(1000000 ether);
     ISablier public sablier;
```
Comment (zgorizzo69) [inline contracts/Bonding.sol:29]: can be removed not needed anymore

File: contracts/Bonding.sol [hunk 3]
```diff
@@ -97,85 +106,138 @@ contract Bonding is CollectableDust {
...
-
-            sablier.createStream(
-                msg.sender,
-                tokenAmount,
-                manager.uADTokenAddress(),
-                streamStart,
-                streamStop
-            );
-        }
+        // console.log("_sharesAmount", _sharesAmount);
```
Comment (zgorizzo69) [inline contracts/Bonding.sol:181]: remove all unnecessary comments like line 244-245

File: contracts/Bonding.sol [hunk 4]
```diff
@@ -97,85 +106,138 @@ contract Bonding is CollectableDust {
...
-        );
+        // for each id of msg.sender
+        // redeemShares(
+        //     IERC20(manager.bondingShareAddress()).balanceOf(msg.sender),
+        //     id
+        // );
     }
 
-    function currentShareValue() public view returns (uint256 pricePerShare) {
+    // SI totalShares = 0  priceShare = TARGET_PRICE
```
Comment (zgorizzo69) [inline contracts/Bonding.sol:227]: translate to EN

Comment (zgorizzo69) [reply contracts/Bonding.sol:227]: can you provide a more literal description for the formulas and use Doxygen-style tags e.g
 /// @dev
  /// @param
   /// @returns
    /// @notice

**Review by zgorizzo69 — COMMENTED** [review]
can we have a separate test file only for UbiquityFormulas unit testing ?

File: contracts/Bonding.sol (hunk 1 above)
Comment (zapaz) [reply contracts/Bonding.sol:125]: ok

File: contracts/Bonding.sol (hunk 2 above)
Comment (zapaz) [reply contracts/Bonding.sol:29]: ok

File: contracts/Bonding.sol (hunk 3 above)
Comment (zapaz) [reply contracts/Bonding.sol:181]: ok line 181+ and 244+

I left sablier stuff commented for later ?

**Pull request closed without merging**